ELEC-71: Telemetry - checksum utils

neilparikh opened the pull request:
This PR adds a stub telemetry project, as well as some utilities for computing checksums of NMEA messages.

tituschow commented on projects/telemetry/inc/util.h:
```diff
@@ -0,0 +1,17 @@
+// and continues until the `*` character.
+//
+// `out` must be a pointer to at least 3 `char`s.
+void evm_gps_compute_checksum(char *message, char *out);
```
Please fix your naming scheme to match the standard and make sure that your inputs take length parameters as well in case they're malformed.

tituschow commented on projects/telemetry/src/util.c:
```diff
@@ -0,0 +1,34 @@
+// out must be a pointer to at least 3 chars
+void prv_int_to_hex(uint8_t checksum, char *out) {
+  // we know that it'll be a 2 digit hex number since the checksum is 8-bit
+  out[0] = s_hex[checksum / 16];
```
I think you could just shift and mask these.

tituschow commented on projects/telemetry/src/util.c:
```diff
@@ -0,0 +1,34 @@
+
+  char computed[3];
+  evm_gps_compute_checksum(message, computed);
+  return strcmp(computed, received) == 0;
```
Is there a reason you don't convert the received checksum into an integer and compare it to the calculated value instead?

tituschow commented on projects/telemetry/src/util.c:
```diff
@@ -0,0 +1,34 @@
+}
+
+bool evm_gps_compare_checksum(char *message) {
+  char *received = message + strlen(message) - 2;
```
Let's avoid strlen since we want to be able to handle malformed strings.

tituschow commented on projects/telemetry/inc/nmea.h:
```diff
@@ -0,0 +1,19 @@
+// are processed, whichever comes first.
+//
+// `out` must be a pointer to at least 3 `char`s.
+void nmea_compute_checksum(char *message, size_t message_len, char *out);
```
Rather than just telling the user through a comment, why not enforce it in your API? You can return a StatusCode.

tituschow replied:
`void nmea_compute_checksum(char *message, size_t message_len, char *out, size_t out_len)` or something similar.

tituschow commented on projects/telemetry/src/nmea.c:
```diff
@@ -0,0 +1,33 @@
+void prv_int_to_hex(uint8_t checksum, char *out) {
+  // we know that it'll be a 2 digit hex number since the checksum is 8-bit
+  out[0] = s_hex[checksum >> 4];
+  out[1] = s_hex[checksum & 15];
```
Can you use hex instead? Using decimal values for bitwise operations feels a little odd. It's hard to translate in your head.

tituschow replied:
Yes, that would be easier to understand.

tituschow commented on projects/telemetry/src/nmea.c:
```diff
@@ -0,0 +1,33 @@
+
+void nmea_compute_checksum(char *message, size_t message_len, char *out) {
+  uint8_t sum = 0;
+  for (uint8_t i = 1; message[i] != '*' && i < message_len; i++) {
```
You skip the first character?

tituschow replied:
Please add a comment for that.

tituschow commented on projects/telemetry/inc/nmea.h:
```diff
@@ -0,0 +1,19 @@
+// checksum against the checksum in the message.
+//
+// Returns false if the given message does not have a checksum included.
+bool nmea_compare_checksum(char *message, size_t message_len);
```
One thing about the `message_len` parameters - Usually, these length args refer to the number of bytes in the array, not the string length. Is that your intention as well?

tituschow replied:
Please make that clear. Note that none of our other modules use that (including UART), so I would suggest switching it for consistency.

neilparikh replied:
Okay, I'll switch to number of bytes, and add that to the comment.

tituschow commented on projects/telemetry/src/nmea.c:
```diff
@@ -0,0 +1,33 @@
+}
+
+bool nmea_compare_checksum(char *message, size_t message_len) {
+  char *received = message + message_len - 2;
```
Could you explain what this is doing? I'm not sure why you're getting the pointer of a near-end value and then subtracting 1 from the pointer below before dereferencing it. It seems a little roundabout and difficult to understand.

tituschow replied:
It might help if you had some higher level explanations of what this checksum is doing and how it fits into the message format in your header.

tituschow commented on projects/telemetry/test/test_nmea.c:
```diff
@@ -0,0 +1,42 @@
+      "M,15.2,M,,0000*64";
+
+  size_t message_len = strlen(input);
+  TEST_ASSERT_TRUE(nmea_compare_checksum(input, message_len));
```
You should try testing some edge cases too - empty string, malformed string, incorrect checksum, etc.

tituschow commented on projects/telemetry/src/nmea.c:
```diff
@@ -0,0 +1,33 @@
+  for (uint8_t i = 1; message[i] != '*' && i < message_len; i++) {
+    sum ^= message[i];
+  }
+  prv_int_to_hex(sum, out);
```
I know you mentioned that the reason the checksum function returns a string is for when you're building messages to send to the GPS, but is there a reason you don't just use `snprintf`? I assume you need to use something like it for building the rest of your command anyways.

tituschow replied:
I meant to use `snprintf` in your command building and just use integers here.

neilparikh replied:
I could return an int here, but I would need to convert the int to hex when I'm building the command anyway, and in `compare_checksum`, I would need to either convert the int to hex, or convert the hex to int, so this seems like the best approach.

What I could do is return an int here, and then call `prv_int_to_hex` in `compare_checksum` instead, if that works.

tituschow commented:
Just a few minor comments and I think this will be good to go

neilparikh commented:
@tituschow: addressed all comments except the one about return an `int` from `nmea_compute_checksum`, which IMO isn't worth fixing until there are a few more users of that function, which will provide a better idea of which format is needed the most.

tituschow commented on projects/telemetry/inc/nmea.h:
```diff
@@ -0,0 +1,33 @@
+// An example NMEA message is "$GPGLL,2503.6319,N,12136.0099,E,053740.000,A,A*52"
+// In this message, the checksum would be 52
+
+// Computes the checksum for a given NMEA message.
```
We usually keep function comments in one block. Right now, it's hard to tell where your description of the message format ends and the function comments start.

tituschow commented on projects/telemetry/src/nmea.c:
```diff
@@ -0,0 +1,45 @@
+  // We expect the last 2 characters in the message to be the sent checksum
+  // Hence, the 3rd last character should be '*' (we subtract 3 instead of 2 to
+  // account for \0)
+  char *received = message + message_len - 3;
```
You should probably check for a minimum message length

tituschow commented on projects/telemetry/inc/nmea.h:
```diff
@@ -0,0 +1,33 @@
+#pragma once
+#include <stdbool.h>
```
We usually try to have a high-level module description here.

mikzharov replied:
Got it, thanks

tituschow commented on .gitignore:
```diff
@@ -36,6 +36,7 @@
 
 # Other
 *.vscode
+.vs/
```
VSCode uses the `.vscode` directory to store its metadata. Are you sure this is actually used?

mikzharov replied:
Yup, it's storing various files like `VSWorkspaceState.json` and  `ProjectSettings.json`

tituschow replied:
VSCode? Or Visual Studio? Because VSCode's metadata is stored in `.vscode`.

tituschow commented on projects/telemetry/inc/nmea.h:
```diff
@@ -0,0 +1,25 @@
+// followed by a *, and then the message checksum (which is a 2-digit hex number).
+// This header will aid in parsing the NMEA messages sent by the GPS.
+
+#pragma once
```
By convention `#pragma once` should be the first line. The usual format is:

```
#pragma once
// module summary
// requires ...
// module comments/description
#include ...
```

mikzharov replied:
Thanks for the heads up, moved it

tituschow commented on projects/telemetry/inc/nmea.h:
```diff
@@ -0,0 +1,25 @@
+// # The NMEA message format
```
Is this module all that the NMEA format is associated with? If there's more to the format, I might recommend remaining this to `nmea_checksum` or something similar.

mikzharov replied:
Ok, done

tituschow commented on projects/telemetry/inc/nmea.h:
```diff
@@ -0,0 +1,25 @@
+// are processed, whichever comes first.
+// * message_len is byte size of message
+// `out` must be a pointer to at least 3 `char`s.
+StatusCode nmea_compute_checksum(char *message, size_t message_len, char *out, size_t out_len);
```
This should probably just output an integer instead - it's a little odd to output a string for a checksum.

tituschow commented on projects/telemetry/src/nmea.c:
```diff
@@ -0,0 +1,52 @@
+// In the message above, the checksum would be 52.
+StatusCode nmea_compute_checksum(char *message, size_t message_len, char *out, size_t out_len) {
+  if (out_len < 3) {
+    return STATUS_CODE_INVALID_ARGS;
```
Remember to use `status_code` or `status_msg` for non-OK returns

tituschow commented on projects/telemetry/inc/nmea_checksum.h:
```diff
@@ -0,0 +1,23 @@
+// Starts at the second characer of `message`,
+// and continues until the `*` character, or until `message_len` characters
+// are processed, whichever comes first.
+StatusCode nmea_compute_checksum(char *message, size_t message_len, uint8_t *out);
```
Remember to update your functions for the module name

tituschow commented on projects/telemetry/inc/nmea_checksum.h:
```diff
@@ -0,0 +1,23 @@
+// * message_len is byte size of message
+// * Returns false if the given message does not have a checksum included or if the
+//   checksums do not match.
+bool nmea_compare_checksum(char *message, size_t message_len);
```
Maybe `validate` instead of `compare`?

mikzharov replied:
Fixed

mikzharov commented on projects/telemetry/src/nmea_checksum.c:
```diff
@@ -0,0 +1,51 @@
+// In the message above, the checksum would be 52.
+StatusCode nmea_compute_checksum(char *message, size_t message_len, uint8_t *out) {
+  if (out == NULL || message == NULL || message_len < 4) {
+    return STATUS_CODE_INVALID_ARGS;
```
Ok, fixed

tituschow commented on .travis.yml:
```diff
@@ -19,6 +19,7 @@ branches:
 
 addons:
   apt:
+    update: true
```
Turns out you don't need this - it was just an availability thing.

mikzharov replied:
Ok

tituschow commented on projects/telemetry/src/nmea_checksum.c:
```diff
@@ -37,12 +39,11 @@ bool nmea_compare_checksum(char *message, size_t message_len) {
   // Extracts the received checksum
-  uint32_t received_checksum = 0;
-  sscanf(received, "%2x", (unsigned int *)&received_checksum);
+  uint32_t received_checksum = (uint32_t)strtol(received, NULL, 16);
```
If you're going to use `strtol`, why not use `strtoul`? Alternatively, you could write a simple hex to int function. Since your checksum should always be 2 hex characters, your checksum should always be a `uint8_t`.

tituschow commented on projects/telemetry/test/test_nmea_checksum.c:
```diff
@@ -7,65 +7,65 @@
 void teardown_test(void) {}
 
-void test_compute_checksum(void) {
+void test_checksum_compute(void) {
```
`test_nmea_checksum_[x]`

tituschow commented on projects/telemetry/src/nmea_checksum.c:
```diff
@@ -0,0 +1,79 @@
+    return h - 'A' + 10;
+  }
+  // Just to make the checksum fail
+  return 16;
```
Please use a define instead of magic numbers

tituschow commented on projects/telemetry/src/nmea_checksum.c:
```diff
@@ -0,0 +1,79 @@
+  // We expect the last 2 characters in the message to be the sent checksum
+  // Hence, the 3rd last character should be '*' (we subtract 3 instead of 2 to
+  // account for \0)
+  char *received = message + message_len - 3;
```
You might as well just treat it as a char array and index into message instead of manual pointer math.

tituschow commented on projects/telemetry/src/nmea_checksum.c:
```diff
@@ -0,0 +1,79 @@
+  // Computes the checksum
+  uint8_t computed = 0;
+  StatusCode status_computed = nmea_checksum_compute(message, message_len, &computed);
+  if (status_computed != STATUS_CODE_OK || status_extracted != STATUS_CODE_OK) {
```
Please use `status_ok`

tituschow commented on projects/telemetry/src/nmea_checksum.c:
```diff
@@ -0,0 +1,79 @@
+    return STATUS_CODE_INVALID_ARGS;
+  }
+  if (computed != NULL) {
+    *computed = int_tens * 16 + int_ones;
```
Just so you know, this is equivalent to `((x & 0xF) << 4) | (y & 0xF)`. (masks are mostly for safety)

tituschow commented on projects/telemetry/src/nmea_checksum.c:
```diff
@@ -0,0 +1,82 @@
+#include <stdlib.h>
+#include <string.h>
+
+static const int HEX_ERROR = 16;
```
Use a #define instead of a variable. Also, make sure you're following the naming scheme.

tituschow commented on projects/telemetry/src/nmea_checksum.c:
```diff
@@ -0,0 +1,82 @@
+    return STATUS_CODE_INVALID_ARGS;
+  }
+  if (computed != NULL) {
+    // Bitwise operation is equivilent to "int_tens * 16 + int_ones"
```
s/equivilent/equivalent/

tituschow commented on projects/telemetry/src/nmea_checksum.c:
```diff
@@ -0,0 +1,82 @@
+
+  // Checks for the 16 because it is the result sent in case of error
+  if (ones == HEX_ERROR || tens == HEX_ERROR) {
+    return STATUS_CODE_INVALID_ARGS;
```
`status_code` or `status_msg`.

tituschow commented on projects/telemetry/src/nmea_checksum.c:
```diff
@@ -0,0 +1,82 @@
+  }
+  if (computed != NULL) {
+    // Bitwise operation is equivalent to "int_tens * 16 + int_ones"
+    *computed = ((int_tens & 0xF) << 4) | (int_ones & 0xF);
```
Just FYI: The compiler probably would've optimized it into the same code.

Your comment isn't actually fully accurate, as the bitwise approach is more "correct" since it masks the values. It's only because you always limit both variables to `[0x00, 0x0F]` that they're equivalent, and if you trust that will always be the case, you can actually remove the masks. Just something to keep in mind.

tituschow commented on projects/telemetry/src/nmea_checksum.c:
```diff
@@ -0,0 +1,82 @@
+#include <stdlib.h>
+#include <string.h>
+
+#define NMEA_CHECKSUM_HEX_ERROR 16
```
Should probably add a comment mentioning why this is valid as an error. Since you're dealing with hex it might be easier to understand to use `0x[y]0` where `y` is a valid hex digit.
